nrf-mpsl/src/flash.rs

```diff
@@ -0,0 +1,496 @@
+        dest: *mut u32,
+        /// Source address
+        src: *const u32,
+        words: u32,
+    },
+}
+
+// Safety:
+// Timeslot request and return parameters are only modified before and after timeslot starts,
+// or within the callback.
+unsafe impl Send for State {}
+unsafe impl Sync for State {}
+
+// Values derived from nRF SDK
+const TIMESLOT_TIMEOUT_PRIORITY_NORMAL_US: u32 = 30000;
+// Extra slack for non-flash activity
+const TIMESLOT_SLACK_US: u32 = 1000;
+
+// Values according to nRF52 product specification
+const ERASE_PAGE_DURATION_US: u32 = 85_000;
+const WRITE_WORD_DURATION_US: u32 = 41;
+const ERASE_PARTIAL_PAGE_DURATION_MS: u32 = 10;
+const ERASE_PARTIAL_PAGE_DURATION_US: u32 = ERASE_PARTIAL_PAGE_DURATION_MS * 1000;
+const WORD_SIZE: u32 = 4;
+
+// Values derived from Zephyr
+const TIMESLOT_LENGTH_ERASE_US: u32 = ERASE_PARTIAL_PAGE_DURATION_US;
+const TIMESLOT_LENGTH_WRITE_US: u32 = 7500;
+
+static STATE: State = State::new();
+
+impl<'d> Flash<'d> {
+    /// Constructs a flash instance.
+    ///
+    /// # Panics
+    ///
+    /// Panics if called more than once.
+    pub fn take(_mpsl: &'d MultiprotocolServiceLayer<'d>, _p: impl Peripheral<P = NVMC> + 'd) -> Flash<'d> {
+        STATE.with_inner(|state| {
+            if state.taken {
+                panic!("nrf_mpsl::Flash::take() called multiple times.")
+            }
+            state.taken = true;
+        });
+
+        into_ref!(_p);
+
+        Self { _mpsl, _p }
+    }
+
+    pub fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), FlashError> {
+        if offset as usize >= FLASH_SIZE || offset as usize + bytes.len() > FLASH_SIZE {
+            return Err(FlashError::OutOfBounds);
+        }
+
+        let flash_data = unsafe { slice::from_raw_parts(offset as *const u8, bytes.len()) };
+        bytes.copy_from_slice(flash_data);
+        Ok(())
+    }
+
+    // Perform a flash operation using the timeslot API
+    //
+    // This method creates a timeslot session, reserves a timeslot to run the flash operation,
+    // and waits until the operation is done. Any result associated with the operation is propagated
+    // to the caller.
+    //
+    // If the future is cancelled, or the operation completes, the session is closed.
+    async fn do_op(&mut self, slot_duration_us: u32, op: FlashOp) -> Result<(), FlashError> {
+        let session_id: u8 = poll_fn(|cx| {
+            STATE.with_inner(|state| {
+                if let FlashOp::None = state.operation {
+                    let mut session_id: u8 = 0;
+                    let ret = unsafe { raw::mpsl_timeslot_session_open(Some(timeslot_session_callback), (&mut session_id) as *mut _) };
+                    match RetVal::from(ret).to_result() {
+                        Ok(_) => {
+                            Poll::Ready(Ok(session_id))
+                        }
+                        Err(Error::ENOMEM) => {
+                            state.waker.register(cx.waker());
```

Comment:
The waker needs to be registered before checking the condition (the call to `mpsl_timeslot_session_open`) to avoid a race condition.

If another timeslot user has taken all the available sessions, I don't think this will ever become ready. It might be better to use `state.operation == FlashOp::None` as the wait condition.

You may also need another waker here, because if a flash operation is in progress, `state.waker` needs to wake the flash completion future below.

Reply:
I don't think there is a race in this case, because the state is protected by a critical section mutex, so the callback wouldn't call the waker before it could lock the mutex. But I can add the registering in case we refactor this later and expose a race.

The reason for using the session open as a waiting condition is that after the previous operation is done, the session id is not free'd until the callback has run. But I can change it so that the callback actually resets state.operation, which should allow it to be used as the condition.

Not sure about the in progress waker, will check it.

Reply:
> I don't think there is a race in this case, because the state is protected by a critical section mutex, so the callback wouldn't call the waker before it could lock the mutex. But I can add the registering in case we refactor this later and expose a race.

Hmm good point, but that actually makes me think that the `CritialSectionRawMutex` is unsound in this case, because the MPSL critical section impl does not mask the `TIMER0` interrupt and the timeslot callback executed in the `TIMER0` interrupt context.

> The reason for using the session open as a waiting condition is that after the previous operation is done, the session id is not free'd until the callback has run. But I can change it so that the callback actually resets state.operation, which should allow it to be used as the condition.

Yeah, I think maybe we need to wait to set operation to `FlashOp::None` until the session closed indication is received? The issue I'm worried about is that if some other code has consume the available timeslot sessions, or if no timeslot sessions were allocated initially, right now this will wait forever instead of returning an `ENOMEM` error.

Reply:
That's a very good point, I forgot that TIMER0 was excluded from that. I will need to rewrite this part a bit then.

Reply:
I've added a custom RawMutex implementation now that only masks timer0.